src/ring_theory/witt_vector/truncated.lean

```diff
@@ -0,0 +1,246 @@
+(`truncated_witt_vector p₁ n R` and `truncated_witt_vector p₂ n R` are definitionally
+equal as types but will have different ring operations.)
+-/
+@[nolint unused_arguments]
+def truncated_witt_vector (p : ℕ) (n : ℕ) (R : Type*) := fin n → R
+
+instance (p n : ℕ) (R : Type*) [inhabited R] : inhabited (truncated_witt_vector p n R) :=
+⟨λ _, default R⟩
+
+variables {n R}
+
+namespace truncated_witt_vector
+
+variables (p)
+
+/-- Create a `truncated_witt_vector` from a vector `x`. -/
+def mk (x : fin n → R) : truncated_witt_vector p n R := x
+
+variables {p}
+
+/-- `x.coeff i` is the `i`th entry of `x`. -/
+def coeff (i : fin n) (x : truncated_witt_vector p n R) : R := x i
+
+@[ext]
+lemma ext {x y : truncated_witt_vector p n R} (h : ∀ i, x.coeff i = y.coeff i) : x = y :=
+funext $ λ n, h n
+
+lemma ext_iff {x y : truncated_witt_vector p n R} : x = y ↔ ∀ i, x.coeff i = y.coeff i :=
+⟨λ h i, by rw h, ext⟩
+
+@[simp] lemma coeff_mk (x : fin n → R) (i : fin n) :
+  (mk p x).coeff i = x i := rfl
+
+@[simp] lemma mk_coeff (x : truncated_witt_vector p n R) :
+  mk p (λ i, x.coeff i) = x :=
+by { ext i, rw [coeff_mk] }
+
+variable [comm_ring R]
+
+/--
+We can turn a truncated Witt vector `x` into a Witt vector
+by setting all coefficients after `x` to be 0.
+-/
+def out (x : truncated_witt_vector p n R) : 𝕎 R :=
+witt_vector.mk p $ λ i, if h : i < n then x.coeff ⟨i, h⟩ else 0
+
+@[simp]
+lemma coeff_out (x : truncated_witt_vector p n R) (i : fin n) :
+  x.out.coeff i = x.coeff i :=
+by rw [out, witt_vector.coeff_mk, dif_pos i.is_lt, fin.eta]
+
+lemma out_injective : injective (@out p n R _) :=
+begin
+  intros x y h,
+  ext i,
+  rw [witt_vector.ext_iff] at h,
+  simpa only [coeff_out] using h ↑i
+end
+
+end truncated_witt_vector
+
+namespace witt_vector
+
+variables {p} (n)
+
+section
+
+local attribute [semireducible] witt_vector
+
+/-- `truncate_fun n x` uses the first `n` entries of `x` to construct a `truncated_witt_vector`,
+which has the same base `p` as `x`.
+This function is bundled into a ring homomorphism in `witt_vector.truncate` -/
+def truncate_fun (x : 𝕎 R) : truncated_witt_vector p n R :=
+truncated_witt_vector.mk p $ λ i, x.coeff i
+
+end
+
+variables {n}
+
+@[simp] lemma coeff_truncate_fun (x : 𝕎 R) (i : fin n) :
+  (truncate_fun n x).coeff i = x.coeff i :=
+by rw [truncate_fun, truncated_witt_vector.coeff_mk]
+
+variable [comm_ring R]
+
+@[simp] lemma out_truncate_fun (x : 𝕎 R) :
+  (truncate_fun n x).out = init n x :=
+begin
+  ext i,
+  dsimp [truncated_witt_vector.out, init, select],
+  split_ifs with hi, swap, refl,
+  rw [coeff_truncate_fun, fin.coe_mk],
+end
+
+end witt_vector
+
+namespace truncated_witt_vector
+
+variable [comm_ring R]
+
+@[simp] lemma truncate_fun_out (x : truncated_witt_vector p n R) :
+  x.out.truncate_fun n = x :=
+by simp only [witt_vector.truncate_fun, coeff_out, mk_coeff]
+
+end truncated_witt_vector
+
+namespace truncated_witt_vector
+open witt_vector
+variables (p n R)
+variable [comm_ring R]
```

Comment:
I checked #5163 too and I didn't see an obvious reason to exit and then enter the same namespace.
```suggestion
open witt_vector
variables (p n R)
```